### ISSUE TITLE
RTL8723DE - Ubuntu 20.04 - Hard blocked after suspend

### ISSUE THREAD

**Issue by facekapow** [lwfinger/rtlwifi_new#588]
Hi, first off, I'd like to say thanks for this driver! It works wonderfully on my system, with one hitch: after suspending/sleeping my computer, the wifi no longer works and `rfkill` reports that the device is hard blocked.

My specs:
  * **Laptop** - HP 15t-dy100
  * **Network card** - RTL8723DE
  * **Operating System** - Ubuntu 20.04
  * **Kernel** - 5.4.0-32-generic

I'm using some prerelease software, but I was having this issue even with the 100% stable release of 20.04.

`sudo lspci -v` output (relevant entry; before suspending; not sure if it would change after suspending):
```
01:00.0 Network controller: Realtek Semiconductor Co., Ltd. RTL8723DE 802.11b/g/n PCIe Adapter
	DeviceName: Realtek Wireless LAN + BT
	Subsystem: Hewlett-Packard Company RTL8723DE 802.11b/g/n PCIe Adapter
	Flags: bus master, fast devsel, latency 0, IRQ 147
	I/O ports at 3000 [size=256]
	Memory at 80600000 (64-bit, non-prefetchable) [size=64K]
	Capabilities: [40] Power Management version 3
	Capabilities: [50] MSI: Enable+ Count=1/1 Maskable- 64bit+
	Capabilities: [70] Express Endpoint, MSI 00
	Capabilities: [100] Advanced Error Reporting
	Capabilities: [148] Virtual Channel
	Capabilities: [168] Device Serial Number 00-e0-4c-00-00-00-00-00
	Capabilities: [178] Latency Tolerance Reporting
	Capabilities: [180] L1 PM Substates
	Kernel driver in use: rtl8723de
	Kernel modules: rtwpci, rtl8723de
```

My modprobe config for `rtl8723de`:
```
options rtl8723de ant_sel=2
```

I've also tried disable power management (`ips=0 fwlps=0`) but had the same result.

`rfkill` output before suspending:
```
ID TYPE      DEVICE      SOFT      HARD
 0 bluetooth hci0   unblocked unblocked
 1 wlan      phy0   unblocked unblocked
```

`rfkill` output after suspending:
```
ID TYPE DEVICE      SOFT    HARD
 1 wlan phy0   unblocked blocked
```

Note that after suspending, my bluetooth is also gone (not sure if that's related). That's not so much of a problem for me at the moment, but the wifi being blocked is.

`dmesg` output after suspending (note that I've only included the output that wasn't there before suspending):
```
[ 6568.725861] PM: suspend entry (deep)
[ 6568.732747] Filesystems sync: 0.006 seconds
[ 6568.733128] Freezing user space processes ... (elapsed 0.002 seconds) done.
[ 6568.735341] OOM killer disabled.
[ 6568.735341] Freezing remaining freezable tasks ... (elapsed 0.001 seconds) done.
[ 6568.736739] printk: Suspending console(s) (use no_console_suspend to debug)
[ 6568.737011] wlo1: deauthenticating from 60:38:e0:df:59:ed by local choice (Reason: 3=DEAUTH_LEAVING)
[ 6568.954102] ACPI: EC: interrupt blocked
[ 6569.005157] ACPI: Preparing to enter system sleep state S3
[ 6569.020026] ACPI: EC: event blocked
[ 6569.020028] ACPI: EC: EC stopped
[ 6569.020028] PM: Saving platform NVS memory
[ 6569.020041] Disabling non-boot CPUs ...
[ 6569.021798] smpboot: CPU 1 is now offline
[ 6569.024725] IRQ 129: no longer affine to CPU2
[ 6569.025756] smpboot: CPU 2 is now offline
[ 6569.029213] smpboot: CPU 3 is now offline
[ 6569.031351] IRQ 126: no longer affine to CPU4
[ 6569.032608] smpboot: CPU 4 is now offline
[ 6569.034697] IRQ 122: no longer affine to CPU5
[ 6569.034715] IRQ 147: no longer affine to CPU5
[ 6569.035764] smpboot: CPU 5 is now offline
[ 6569.037939] IRQ 123: no longer affine to CPU6
[ 6569.038977] smpboot: CPU 6 is now offline
[ 6569.041375] IRQ 17: no longer affine to CPU7
[ 6569.041393] IRQ 125: no longer affine to CPU7
[ 6569.042428] smpboot: CPU 7 is now offline
[ 6569.047620] ACPI: Low-level resume complete
[ 6569.047713] ACPI: EC: EC started
[ 6569.047713] PM: Restoring platform NVS memory
[ 6569.048592] Enabling non-boot CPUs ...
[ 6569.048637] x86: Booting SMP configuration:
[ 6569.048637] smpboot: Booting Node 0 Processor 1 APIC 0x2
[ 6569.049563] CPU1 is up
[ 6569.049588] smpboot: Booting Node 0 Processor 2 APIC 0x4
[ 6569.050531] CPU2 is up
[ 6569.050573] smpboot: Booting Node 0 Processor 3 APIC 0x6
[ 6569.051579] CPU3 is up
[ 6569.051615] smpboot: Booting Node 0 Processor 4 APIC 0x1
[ 6569.052643] CPU4 is up
[ 6569.052672] smpboot: Booting Node 0 Processor 5 APIC 0x3
[ 6569.053834] CPU5 is up
[ 6569.053877] smpboot: Booting Node 0 Processor 6 APIC 0x5
[ 6569.054996] CPU6 is up
[ 6569.055036] smpboot: Booting Node 0 Processor 7 APIC 0x7
[ 6569.055975] CPU7 is up
[ 6569.059082] ACPI: Waking up from system sleep state S3
[ 6569.068911] ACPI: EC: interrupt unblocked
[ 6569.120631] ACPI: EC: event unblocked
[ 6569.122500] rtlwifi: rtlwifi: wireless switch is on
[ 6569.129895] i2c_hid i2c-ELAN0712:00: i2c_hid_get_input: IRQ triggered but there's no data
[ 6569.130956] i2c_hid i2c-ELAN0712:00: i2c_hid_get_input: incomplete report (14/3583)
[ 6569.132940] i2c_hid i2c-ELAN0712:00: i2c_hid_get_input: incomplete report (14/65280)
[ 6569.237714] nvme nvme0: 8/0/0 default/read/poll queues
[ 6569.242196] nvme nvme1: 8/0/0 default/read/poll queues
[ 6569.411821] usb 1-3: reset high-speed USB device number 2 using xhci_hcd
[ 6572.087565] acpi LNXPOWER:05: Turning OFF
[ 6572.088327] mei_hdcp 0000:00:16.0-b638ab7e-94e2-4ea2-a552-d1c54b627f04: bound 0000:00:02.0 (ops i915_hdcp_component_ops [i915])
[ 6572.091490] acpi LNXPOWER:04: Turning OFF
[ 6572.094779] acpi LNXPOWER:03: Turning OFF
[ 6572.094802] OOM killer enabled.
[ 6572.094802] Restarting tasks ... done.
[ 6572.120251] usb 1-10: USB disconnect, device number 3
[ 6572.183504] PM: suspend exit
```

Loaded modules:
```
Module                  Size  Used by
rfcomm                 81920  0
ccm                    20480  0
cmac                   16384  0
algif_hash             16384  0
algif_skcipher         16384  0
af_alg                 24576  2 algif_hash,algif_skcipher
bnep                   24576  2
snd_hda_codec_hdmi     61440  1
snd_sof_pci            20480  0
snd_sof_intel_hda_common    69632  1 snd_sof_pci
snd_soc_hdac_hda       24576  1 snd_sof_intel_hda_common
snd_sof_intel_hda      20480  1 snd_sof_intel_hda_common
snd_sof_intel_byt      20480  1 snd_sof_pci
snd_sof_intel_ipc      20480  1 snd_sof_intel_byt
snd_sof               106496  4 snd_sof_pci,snd_sof_intel_hda_common,snd_sof_intel_byt,snd_sof_intel_ipc
snd_sof_xtensa_dsp     16384  1 snd_sof_pci
snd_hda_ext_core       28672  3 snd_sof_intel_hda_common,snd_soc_hdac_hda,snd_sof_intel_hda
snd_soc_acpi_intel_match    32768  2 snd_sof_pci,snd_sof_intel_hda_common
snd_soc_acpi           16384  2 snd_sof_pci,snd_soc_acpi_intel_match
snd_soc_core          245760  3 snd_sof,snd_sof_intel_hda_common,snd_soc_hdac_hda
snd_hda_codec_realtek   118784  1
snd_hda_codec_generic    81920  1 snd_hda_codec_realtek
ledtrig_audio          16384  3 snd_hda_codec_generic,snd_hda_codec_realtek,snd_sof
snd_compress           24576  1 snd_soc_core
ac97_bus               16384  1 snd_soc_core
snd_pcm_dmaengine      16384  1 snd_soc_core
snd_hda_intel          53248  3
snd_intel_dspcfg       24576  3 snd_hda_intel,snd_sof_pci,snd_sof_intel_hda_common
snd_hda_codec         131072  5 snd_hda_codec_generic,snd_hda_codec_hdmi,snd_hda_intel,snd_hda_codec_realtek,snd_soc_hdac_hda
snd_hda_core           90112  9 snd_hda_codec_generic,snd_hda_codec_hdmi,snd_hda_intel,snd_hda_ext_core,snd_hda_codec,snd_hda_codec_realtek,snd_sof_intel_hda_common,snd_soc_hdac_hda,snd_sof_intel_hda
snd_hwdep              20480  1 snd_hda_codec
snd_pcm               106496  8 snd_hda_codec_hdmi,snd_hda_intel,snd_hda_codec,snd_sof,snd_sof_intel_hda_common,snd_soc_core,snd_hda_core,snd_pcm_dmaengine
snd_seq_midi           20480  0
snd_seq_midi_event     16384  1 snd_seq_midi
mei_hdcp               24576  0
intel_rapl_msr         20480  0
nls_iso8859_1          16384  1
rtl8723de              90112  0
btcoexist             352256  1 rtl8723de
x86_pkg_temp_thermal    20480  0
snd_rawmidi            36864  1 snd_seq_midi
intel_powerclamp       20480  0
phydm_mod             753664  1 rtl8723de
coretemp               20480  0
snd_seq                69632  2 snd_seq_midi,snd_seq_midi_event
joydev                 24576  0
uvcvideo               98304  0
videobuf2_vmalloc      20480  1 uvcvideo
rtl8723_common         24576  1 rtl8723de
snd_seq_device         16384  3 snd_seq,snd_seq_midi,snd_rawmidi
videobuf2_memops       20480  1 videobuf2_vmalloc
rtl_pci                28672  1 rtl8723de
snd_timer              36864  2 snd_seq,snd_pcm
rtlwifi                94208  5 rtl_pci,rtl8723de,btcoexist,phydm_mod,rtl8723_common
videobuf2_v4l2         24576  1 uvcvideo
kvm_intel             282624  0
videobuf2_common       49152  2 videobuf2_v4l2,uvcvideo
mac80211              843776  2 rtl_pci,rtlwifi
kvm                   663552  1 kvm_intel
videodev              225280  3 videobuf2_v4l2,uvcvideo,videobuf2_common
btusb                  57344  0
btrtl                  24576  1 btusb
intel_cstate           20480  0
btbcm                  16384  1 btusb
btintel                24576  1 btusb
input_leds             16384  0
hp_wmi                 16384  0
intel_rapl_perf        20480  0
serio_raw              20480  0
sparse_keymap          16384  1 hp_wmi
mc                     53248  4 videodev,videobuf2_v4l2,uvcvideo,videobuf2_common
snd                    90112  19 snd_hda_codec_generic,snd_seq,snd_seq_device,snd_hda_codec_hdmi,snd_hwdep,snd_hda_intel,snd_hda_codec,snd_hda_codec_realtek,snd_timer,snd_compress,snd_soc_core,snd_pcm,snd_rawmidi
bluetooth             581632  14 btrtl,btintel,btbcm,bnep,btusb,rfcomm
wmi_bmof               16384  0
hid_multitouch         28672  0
cfg80211              704512  2 rtlwifi,mac80211
soundcore              16384  1 snd
ecdh_generic           16384  1 bluetooth
ecc                    28672  1 ecdh_generic
processor_thermal_device    24576  0
mei_me                 40960  1
libarc4                16384  1 mac80211
mei                   106496  3 mei_hdcp,mei_me
intel_rapl_common      24576  2 intel_rapl_msr,processor_thermal_device
intel_soc_dts_iosf     20480  1 processor_thermal_device
hp_wireless            16384  0
acpi_pad              184320  0
mac_hid                16384  0
int3400_thermal        20480  0
acpi_tad               16384  0
int3403_thermal        16384  0
acpi_thermal_rel       16384  1 int3400_thermal
int340x_thermal_zone    16384  2 int3403_thermal,processor_thermal_device
sch_fq_codel           20480  5
parport_pc             40960  0
ppdev                  24576  0
lp                     20480  0
parport                53248  3 parport_pc,lp,ppdev
ip_tables              32768  0
x_tables               40960  1 ip_tables
autofs4                45056  2
dm_cache_smq           24576  1
dm_cache               61440  2 dm_cache_smq
dm_persistent_data     69632  1 dm_cache
dm_bio_prison          20480  1 dm_cache
dm_bufio               28672  1 dm_persistent_data
libcrc32c              16384  1 dm_persistent_data
dm_crypt               40960  2
hid_generic            16384  0
crct10dif_pclmul       16384  1
crc32_pclmul           16384  0
ghash_clmulni_intel    16384  0
i915                 1986560  22
i2c_algo_bit           16384  1 i915
drm_kms_helper        184320  1 i915
syscopyarea            16384  1 drm_kms_helper
aesni_intel           372736  4
sysfillrect            16384  1 drm_kms_helper
nvme                   49152  4
crypto_simd            16384  1 aesni_intel
sysimgblt              16384  1 drm_kms_helper
cryptd                 24576  4 crypto_simd,ghash_clmulni_intel
fb_sys_fops            16384  1 drm_kms_helper
glue_helper            16384  1 aesni_intel
nvme_core             102400  6 nvme
i2c_i801               32768  0
drm                   491520  10 drm_kms_helper,i915
intel_lpss_pci         20480  0
ahci                   40960  0
intel_lpss             16384  1 intel_lpss_pci
idma64                 20480  0
libahci                32768  1 ahci
virt_dma               20480  1 idma64
i2c_hid                28672  0
wmi                    32768  2 hp_wmi,wmi_bmof
hid                   131072  3 i2c_hid,hid_multitouch,hid_generic
video                  49152  1 i915
```

I've seen in various places online that the issue is most likely related to WMI, but I've tried blacklisting all of my WMI modules (and rebooting of course) and still had no luck. I've also tried blacklisting `hp_wireless` (at the same time the WMI modules were blacklisted) and still nothing.

Ubuntu 20.04 actually already has a preinstalled driver for the RTL8723DE (`rtwpci`), but I was having a similar issue where after suspending the wifi wouldn't work (but there was no indiciation of the device being blocked; `rfkill` said it was unblocked). But now I'm thinking it's the same issue as this, just that the builtin driver doesn't correctly report what's going on. I blacklisted that module and verified that nothing else was controlling my network card before installing `rtlwifi-new`, so it's not an issue with a conflict between drivers.

I've tried replacing the firmwares for rtlwifi with the ones from this repo (but they're actually the same; I checked with `diff` before replacing), but still get the same result.

Removing (via `modprobe -r` *and* `rmmod`) and reloading the module does not fix the issue either.

I see that in `rtl8723de/hw.c` that currently `rtl8723de_suspend` and `rtl8723de_resume` are no-ops. Is it possible that this is related to that? Maybe something needs to be done on resume?

Thank you for any help you can provide.

**Comment by facekapow** [comment]
Just tried the [smlinux driver](https://github.com/smlinux/rtl8723de) and had the same result (actually, same behavior as the builtin driver in Ubuntu 20.04). What's interesting is that with that driver, `dmesg` actually yields some more output. It seems to be functioning according to the output, but it doesn't actually do anything; it mentions the driver powering up and powering down the network card, and scanning for networks, but it doesn't actually do anything.

**Comment by lwfinger** [comment]
As long as you have a hard block, nothing will work. Complain to Ubuntu as it is something in their user space that is causing the block to happen. The driver does not do that.

No, the no-ops in the driver are not the problem. Note that I no longer support this driver - only the one in branch rtw88. I sduggest that you change and blacklist rtl8723de.

**Comment by facekapow** [comment]
I took a look at the rtw88 branch, saw your message about Ubuntu, and decided to try a different distro.

With Arch, the same problem as the builtin driver in Ubuntu 20.04 occurs: no wifi after suspend (but the card is not blocked). In both distros, the WLAN USB interface device is disconnected after suspending/sleeping. I only installed a base Arch system plus NetworkManager and built and installed the driver, and this occurs, so I can’t imagine it’s an issue with the userspace (unless the issue is NetworkManager). Should I open a separate issue for this, or is there somewhere else I should post this issue?

(note that this is for the rtw88 branch that you recommended; but it also occurs with the extended branch)
